contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+
+Another big issue was needing to run entire batches again if they failed. It wasn’t possible to restart from where we left off. Exports couldn’t pause or run simultaneously either. This put a lot of stress on the batch systems. 
+
+Altogether, this created an inconsistent, unreliable export system.
+
+### Opaque
+
+On top of unreliability, we didn’t have a clear picture of what was happening in the export system. Throughout our codebase, we had multiple different work queues like Celery, Kafka, and Graphile each running separately in different contexts.
+
+We didn’t have logs or reports. For example, we didn’t know how much data was being exported or when duplicate runs were happening. When issues happened, we relied on users to tell us. This led to a lot of firefighting and often manual restarts of exports.
+
+### Expensive
+
+The final nail was that exports were expensive. This wasn’t because we were charging a lot for exports – in fact, we weren’t charging anything. It was because they were expensive on the destination side. 
+
+The unreliability led to larger amounts of data processing on the destination side, all of which come with costs. For example, if your export failed 75% through and you needed to reset, you end up paying ingestion fees of 175%. We were also paying for more processing than needed because exports were failing repeatedly.
+
+On top of extra processing time, exports were taking up large amounts of our team’s time. Reliability needed fixing before we worked on performance, and firefighting always focused on reliability. More importantly, our team was spending large amounts of time firefighting and "babysitting" exports. They had to be reactive, rather than proactive.
+
+Simply, we outgrew the export system we put in place, and needed a more reliable, transparent, and cost-effective system if we wanted to scale further.
+
+## Temporal to the rescue
+
+Early in 2023, one of our engineers, [James](/community/profiles/90), opened an [RFC recommending using Temporal at PostHog](https://github.com/PostHog/meta/pull/99). Temporal is a workflow engine abstracting away the details of failure modes, retry logic, and timeouts. This enables developers to build and deploy rock-solid business logic. It is used by big companies with large amounts of data flows – like Uber, Coinbase, Doordash, and Hashicorp.
+
+In Q2, we set a goal of "rock solid-batch processing" and [chose Temporal](https://github.com/PostHog/meta/pull/99) to build it because it helped solve issues like:
+
+- Retry, resume, cancel, and timeout logic per activity and workflow
+- Provides assurances about running exactly-once
+- Visibility into the status of workflows, including errors and logging.
```

Comment:
Temporal also takes care of [scheduling](https://docs.temporal.io/workflows#schedule).

Reply:
Which we leverage heavily.